subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -103,18 +130,28 @@ public boolean isCacheAllowed() {
 
     @Override
     public void cacheIf(final Spec<? super Task> spec) {
+        cacheIf("Task output is not cacheable", spec);
+    }
+
+    @Override
+    public void cacheIf(final String message, final Spec<? super Task> spec) {
         taskMutator.mutate("TaskOutputs.cacheIf(Spec)", new Runnable() {
             public void run() {
-                cacheIfSpec = cacheIfSpec.and(spec);
+                cacheIfSpecs.add(new SpecWithDescription<TaskInternal>(spec, message));
```

Comment:
Do you think adding this API to Specs is worth it? We currently only record reasons for cacheIf Specs which yield false and doNotCacheIf Specs which yield true. That would mean that either `AndSpec` and `OrSpec` would have to gather the descriptions for specs which yield true and which yield false. I would not add that complexity now but revisit it if we also want to report reasons why up to date checks for a task are disabled.
WDYT?